test/libevmasm/Optimiser.cpp

```diff
@@ -1072,6 +1072,76 @@ BOOST_AUTO_TEST_CASE(cse_sub_zero)
 	});
 }
 
+BOOST_AUTO_TEST_CASE(cse_remove_unwanted_masking_of_address)
+{
+	vector<Instruction> ops{
+		Instruction::ADDRESS,
+		Instruction::CALLER,
+		Instruction::ORIGIN,
+		Instruction::COINBASE
+	};
+	for (auto const& op: ops)
+	{
+		checkCSE({
+			u256("0xffffffffffffffffffffffffffffffffffffffff"),
+			op,
+			Instruction::AND
+		}, {
+			op
+		});
+
+		checkCSE({
+			op,
+			u256("0xffffffffffffffffffffffffffffffffffffffff"),
+			Instruction::AND
+		}, {
+			op
+		});
+
+		// do not remove mask for other masking
+		checkCSE({
+			u256(1234),
+			op,
+			Instruction::AND
+		}, {
+			op,
+			u256(1234),
+			Instruction::AND
+		});
+
+		checkCSE({
+			op,
+			u256(1234),
+			Instruction::AND
+		}, {
+			u256(1234),
+			op,
+			Instruction::AND
+		});
+	}
+
+	// leave other opcodes untouched
+	checkCSE({
+		u256("0xffffffffffffffffffffffffffffffffffffffff"),
+		Instruction::BALANCE,
+		Instruction::AND
+	}, {
+		u256("0xffffffffffffffffffffffffffffffffffffffff"),
+		Instruction::BALANCE,
+		Instruction::AND
+	});
+
+	checkCSE({
+		Instruction::BALANCE,
+		u256("0xffffffffffffffffffffffffffffffffffffffff"),
+		Instruction::AND
+	}, {
+		u256("0xffffffffffffffffffffffffffffffffffffffff"),
+		Instruction::SWAP1,
```

Comment:
This is weird.

Reply:
The semantics are the same (balance takes an argument), although it is weird that it generates sub-optimal code in that case. I think the reason the simplification rule is not applied here is because you did not specify the argument for `balance` in the rule. We should perhaps add an assertion somewhere that the arity of the operations in the rules is correct.

Reply:
Ah sorry, the rule for balance is not there at all.

Reply:
Right, `balance` is one of those misnamed ones, because it requires the address, should be called `extbalance`.

However the problem is there it won't check the arity. In the first case `AND` receives only 1 operand, in the second case `SWAP1` applied to a single stack item *and* `AND` receives only 1 operand.

Reply:
It is kind of hard to ensure that with dynamic jumps though :)

Reply:
No, it's fine. The first example, rewritten in functional style, where `x` is the topmost slot before the code starts executing: `and(0xffff..fff, balance(x))`. After CSE ran over it, eliminating the swap:  `and(balance(0xfff..fff), x)`